### PR TITLE
fix: make sharp image processing library an optional dependency

### PR DESCRIPTION
**Pull request by fson**
## Why?

Installing the `sharp` image processing library can fail for a few resons:
- permission problems (possobly the most common issue): it tries to download prebuilt binaries in the (post?)install script which fails if you have only root permissions in the global `node_modules` folder. https://github.com/expo/expo-cli/issues/650 https://github.com/expo/expo-cli/issues/613 https://github.com/expo/expo-cli/issues/591 https://github.com/expo/expo-cli/issues/590
- connectivity problems: the binary download, although not very big (9 MB) can fail when the user has flaky internet connectivity. The prebuild download script perhaps doesn’t retry as much as npm/Yarn does https://github.com/expo/expo-cli/issues/650
- unsupported platforms: https://github.com/expo/expo-cli/issues/605

When sharp installation fails, previously the whole Expo CLI installation has failed.

## How

To make it possible to install Expo CLI without `sharp`, this PR makes `sharp` an optional dependency (via the `sharp-cli` package). Only certain functionality (`expo optimize` and web builds) require `sharp`.

For most users sharp will be installed without problems, and those users will be able to use all features without any additional setup.

Those users who previously had problems installing `sharp` will continue to see warnings when installing but because the dependency is now optional, it won’t fail the whole installation. They will also be able to use Expo CLI. When trying to use functionality that requires `sharp`, Expo CLI will prompt them
to separately install the `sharp-cli` package globally. After `sharp-cli` is installed, those commands can be used.

Screenshot 2019-06-18 at 20 21 33
Screenshot 2019-06-18 at 20 22 23

## Further improvements

In the future, we could make our own distribution of sharp to fix the permissions issues (using the same method we use for ngrok and traveling fastlane). Additionally we could provide JS-only fallbacks (even if they’re slower) for platforms that sharp doesn’t support.

### CODE REVIEW

File: packages/xdl/src/Project.js [hunk 1]
```diff
@@ -2097,13 +2096,14 @@ export async function optimizeAsync(projectRoot: string = './', options: Object
...
+    const optimizedImage = await optimizeImageAsync(image, quality);
 
-    const { size: newSize } = fs.statSync(image);
+    const { size: newSize } = fs.statSync(optimizedImage);
     const amountSaved = prevSize - newSize;
-    if (amountSaved < 0) {
-      // Delete the optimized version and revert changes
-      fs.renameSync(newName, image);
+    if (amountSaved > 0) {
+      await fs.move(image, newName);
```
Comment (ide) [inline packages/xdl/src/Project.js]: Promise.all

Comment (fson) [reply packages/xdl/src/Project.js]: These operations should not happen in parallel – we first move the old file `filename.orig.ext` and then move the optimized image from a temporary folder to its place.

File: packages/image-utils/src/sharp.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    | 'left bottom'
+    | 'left top'
+    | 'entropy'
+    | 'attention';
+  width: number;
+  withoutEnlargement?: boolean;
+};
+
+type Options =
+  | {}
```
[the remark this answers is not in the document]
Comment (fson) [reply packages/image-utils/src/sharp.ts]: TypeScript complained that `{}` is not compatible with `Options`. (The options for the `removeAlpha` operation are `{}`.)

File: packages/image-utils/src/sharp.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  | {
+      [key: string]: boolean | number | string | undefined;
+    };
+
+const requiredCliVersion = optionalDependencies['sharp-cli'];
+
+const SHARP_HELP_PATTERN = /\n\nSpecify --help for available options/g;
+
+export async function sharpAsync(
```
[the remark this answers is not in the document]
Comment (fson) [reply packages/image-utils/src/sharp.ts]: 👍

**Pull request merged**